runtime/vm/criuhelpers.cpp

```diff
@@ -103,4 +105,110 @@ isCheckpointAllowed(J9VMThread *currentThread)
+	if (NULL == newHook) {
+		setNativeOOMHelper(currentThread, vmFuncs);
+	} else {
+		newHook->instanceType = instanceType;
+		newHook->includeSubClass = includeSubClass;
+		newHook->hookFunc = hookFunc;
+		newHook->instanceObjects = pool_new(sizeof(J9ObjectRecord), 0, 0, 0, J9_GET_CALLSITE(), OMRMEM_CATEGORY_VM, POOL_FOR_PORT(vm->portLibrary));
+		if (NULL == newHook->instanceObjects) {
+			setNativeOOMHelper(currentThread, vmFuncs);
+		}
+	}
+}
+
+static BOOLEAN
+iterateInstances(J9JavaVM *vm, void *userData, j9object_t object, BOOLEAN isRestoreHook)
+{
+	BOOLEAN result = TRUE;
+	J9InternalHookRecord *hookRecord = (J9InternalHookRecord*)userData;
+	J9VMThread *currentThread = vm->mainThread;
+
+	if (isRestoreHook) {
+		PORT_ACCESS_FROM_JAVAVM(vm);
+		pool_state walkState = {0};
+		J9ObjectRecord *objectRecord = (J9ObjectRecord*)pool_startDo(hookRecord->instanceObjects, &walkState);
+		while (NULL != objectRecord) {
+			object = objectRecord->object;
+
+			/* assuming java.lang.Random object which is only instance type added into hookRecords so far */
+			j9object_t seedObj = J9VMJAVAUTILRANDOM_SEED(currentThread, object);
+			if (NULL != seedObj) {
+				/* Algorithm used in Random.seedUniquifier() */
+				IDATA current = J9VMJAVAUTILCONCURRENTATOMICATOMICLONG_VALUE(currentThread, seedObj);
+				current = current * 1181783497276652981;
+				current = current ^ j9time_nano_time();
+				Trc_VM_criu_jur_new_longvalue(currentThread, currentThread, seedObj, current);
+				J9VMJAVAUTILCONCURRENTATOMICATOMICLONG_SET_VALUE(currentThread, seedObj, current);
+			}
+			objectRecord = (J9ObjectRecord*)pool_nextDo(&walkState);
+		}
+	} else {
+		J9Class *clazz = J9OBJECT_CLAZZ_VM(vm, object);
+		if ((clazz == hookRecord->instanceType)
+			|| (hookRecord->includeSubClass && isSameOrSuperClassOf(hookRecord->instanceType, clazz))
+		) {
+			J9ObjectRecord *objectRecord = (J9ObjectRecord*)pool_newElement(hookRecord->instanceObjects);
+			if (NULL == objectRecord) {
+				J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+				result = FALSE;
+				setNativeOOMHelper(currentThread, vmFuncs);
+			} else {
+				objectRecord->object = object;
+			}
+		}
+	}
+
+	return result;
+}
+
+void
+initializeCriuHooks(J9JavaVM *vm)
+{
+	J9VMThread *currentThread = vm->mainThread;
+	J9InternalVMFunctions *vmFuncs = vm->internalVMFunctions;
+
+	vm->checkpointState.hookRecords = pool_new(sizeof(J9InternalHookRecord), 0, 0, 0, J9_GET_CALLSITE(), OMRMEM_CATEGORY_VM, POOL_FOR_PORT(vm->portLibrary));
+	if (NULL == vm->checkpointState.hookRecords) {
+		setNativeOOMHelper(currentThread, vmFuncs);
+	} else {
+		J9Class *juRandom = vmFuncs->internalFindKnownClass(currentThread, J9VMCONSTANTPOOL_JAVAUTILRANDOM, J9_FINDKNOWNCLASS_FLAG_EXISTING_ONLY);
+		vmFuncs->addInternalJVMCheckpointHook(currentThread, juRandom, FALSE, iterateInstances);
+	}
+}
+
+BOOLEAN
+iterateCriuHookRecords(J9JavaVM *vm, j9object_t object, BOOLEAN isRestoreHook)
```

Comment:
we need two versions of this, one for the checkpoint and one for the restore hooks